docs/development/design/explicit-builders.rst

```diff
@@ -0,0 +1,178 @@
+- Unexpected dependencies are installed.
+- The wrong docs directory is used.
+- Their configuration file is changed on build time overriding the defaults or adding new things.
+- Some files are auto-generated (like ``index.{rst,md}``).
+
+Currently we are aiming to remove this *magic* behaviour from our build process,
+and educating users to be more explicit about their dependencies and options
+in order to make their builds reproducible.
+
+We are using several feature flags to stop doing this for new projects,
+but with so many flags to check our code starts to be hard to follow.
+Instead we would manage a single feature flag and several classes of builds and environments.
+
+Python Environments
+-------------------
+
+Currently we have two Python environments: Virtualenv and Conda,
+they are in charge of installing requirements into an isolated environment.
+We would need to refactor those classes into two types: the new build, and the old build.
+
+The new Python environments would install only the latest versions of the following dependencies:
+
+Virtualenv
+   - Sphinx or MkDocs
+   - readthedocs-sphinx-ext
+
+Conda
+   - readthedocs-sphinx-ext
+
+Note that for Conda we won't install Sphinx or MkDocs,
+this is to avoid problems like `#3829`_ and `#8096`_.
+
+.. _#3829: https://github.com/readthedocs/readthedocs.org/issues/3829
+.. _#8096: https://github.com/readthedocs/readthedocs.org/issues/8096
+
+Doc builders
+------------
+
+Currently we have two types of doc builders: Sphinx and MkDocs.
+They are in charge of generating the HTML files (or PDF/EPUB) from the source files.
+We would need to refactor those classes into two types: the new build, and the old build.
+
+Both builders create an ``index.{rst,md}`` file with some suggestions if one doesn't exist,
+this is to avoid surprises to users, but sometimes this is done on purpose.
+We could change our default 404 page to include some of these suggestions instead.
+Related to `#1800`_.
+
+.. _#1800: https://github.com/readthedocs/readthedocs.org/issues/1800
+
+The new builders would do the minimal (or none) changes to the user's configuration in order to build their docs.
+
+Sphinx
+~~~~~~
+
+conf.py
+'''''''
+
+We should read the configuration file from the user or default to *one* path,
+and error if it doesn't exist.
+We shouldn't change or override the values from the user's configuration (``conf.py``),
+some are:
+
+- ``_static`` is added to ``html_static_path`` (probably an old setting)
+- ``html_theme`` (we are only setting this if certain condition is met)
+- ``websupport2_*`` (probably old settings)
+- ``html_context`` (more information below)
+- Latex settings
+  (they are used to improve the output for Chinese and Japanese languages, we should write a guide instead)
+
+Sphinx's ``html_context``
+'''''''''''''''''''''''''
+
+We should pass the minimal needed information into the context.
+This is related to :doc:`/development/design/theme-context`.
```

Comment:
Agreed -- though we should also heavily minimize the data to start, treating it basically as an API 👍

Reply:
It seems that this is not decided yet. The document proposes to remove (if possible) passing `html_context` or at least reduce it.

Considering that we are trying to avoid this magic, I'd propose to do a little research to understand if it's possible to not touch this variable from our side and remove it completely since it's a Sphinx-only feature. Also, because we are trying to remove ourselves from the build process.

(Note: I'm considering using an intermediate tool-agnostic `metadata.yaml` file for required context like this in a future builder)

Reply:
It seems that this is not decided yet. The document proposes to remove (if possible) passing `html_context` or at least reduce it.

Considering that we are trying to avoid this magic, I'd propose to do a little research to understand if it's possible to not touch this variable from our side and remove it completely since it's a Sphinx-only feature. Also, because we are trying to remove ourselves from the build process.

(Note: I'm considering using an intermediate tool-agnostic `metadata.yaml` file for required context like this in a future builder)